implementations/rust/ockam/ockam_identity/src/secure_channel/handshake/handshake_worker.rs

```diff
@@ -0,0 +1,314 @@
+            // start the encryptor worker and return the decryptor
+            self.decryptor = Some(self.finalize(context, final_state).await?);
+            self.callback_sender.send(()).await?;
+        };
+        Ok(())
+    }
+}
+
+impl HandshakeWorker {
+    /// Create a new HandshakeWorker with a role of either INITIATOR or RESPONDER
+    #[allow(clippy::too_many_arguments)]
+    pub(crate) async fn create(
+        context: &Context,
+        secure_channels: Arc<SecureChannels>,
+        addresses: Addresses,
+        identifier: IdentityIdentifier,
+        trust_policy: Arc<dyn TrustPolicy>,
+        decryptor_outgoing_access_control: Arc<dyn OutgoingAccessControl>,
+        credentials: Vec<Credential>,
+        trust_context: Option<TrustContext>,
+        remote_route: Option<Route>,
+        timeout: Option<Duration>,
+        role: Role,
+    ) -> Result<Address> {
+        let vault = to_xx_vault(secure_channels.vault());
+        let identities = secure_channels.identities();
+        let state_machine: Box<dyn StateMachine> = if role.is_initiator() {
+            Box::new(
+                InitiatorStateMachine::new(
+                    vault,
+                    identities,
+                    identifier.clone(),
+                    credentials.clone(),
+                    trust_policy.clone(),
+                    trust_context.clone(),
+                )
+                .await?,
+            )
+        } else {
+            Box::new(
+                ResponderStateMachine::new(
+                    vault,
+                    identities,
+                    identifier.clone(),
+                    credentials.clone(),
+                    trust_policy.clone(),
+                    trust_context.clone(),
+                )
+                .await?,
+            )
+        };
+
+        let (mut callback_waiter, callback_sender) = ockam_node::callback::new_callback();
+
+        let worker = Self {
+            secure_channels,
+            callback_sender,
+            state_machine,
+            identifier: identifier.clone(),
+            role: role.clone(),
+            remote_route: remote_route.clone(),
+            addresses: addresses.clone(),
+            decryptor: None,
+        };
+
+        WorkerBuilder::with_mailboxes(
+            Self::create_mailboxes(&addresses, decryptor_outgoing_access_control),
+            worker,
+        )
+        .start(context)
+        .await?;
+
+        let decryptor_remote = addresses.decryptor_remote.clone();
+        debug!(
+            "Starting SecureChannel {} at remote: {}",
+            role.clone(),
+            &decryptor_remote
+        );
+        trace!(
+            "SecureChannel {} started for addresses: {:?}",
+            role.clone(),
+            addresses.clone(),
+        );
+
+        if role.is_initiator() {
+            if let Some(timeout) = timeout {
+                callback_waiter.receive_timeout(timeout).await?;
```

Comment:
Also that depending on a role this function exits on very different conditions.

Reply:
Also that depending on a role this function exits on very different conditions.

Reply:
> This is unexpected for me that HandshakeWorker::create() would wait for the handshake to complete

If you don't wait you can't start sending messages right away because they won't reach the `EncryptorWorker`

> Also that depending on a role this function exits on very different conditions

I agree. I also think that the part about waiting for the decryptor worker is not even working. Because at the end of the Handshake there's no registered decryptor worker. 

I replaced that with a waiting condition only for the initiator, so that it knows when to send messages:

 - wait for the handshake to finish
 - wait for the encryptor worker to be ready

On the responder side the situation is less problematic since a message to be decrypted will always be queued after the last message in the handshake and will be processed by the decryptor handler which is not a worker (so there's nothing to synchronize)